cpp/include/cugraph/mtmg/detail/per_device_edgelist.hpp

```diff
@@ -110,45 +125,67 @@ class per_device_edgelist_t {
               std::optional<raft::host_span<edge_t const>> edge_id,
               std::optional<raft::host_span<edge_type_t const>> edge_type)
   {
-    // FIXME:  This lock guard could be on a smaller region, but it
-    //   would require more careful coding.  The raft::update_device
-    //   calls could be done without the lock if we made a local
-    //   of the values of *.back() and did an increment of current_pos_
-    //   while we hold the lock.
-    std::lock_guard<std::mutex> lock(lock_);
+    std::vector<std::tuple<vertex_t*, vertex_t const*, size_t>> src_copies;
+    std::vector<std::tuple<vertex_t*, vertex_t const*, size_t>> dst_copies;
+    std::vector<std::tuple<weight_t*, weight_t const*, size_t>> wgt_copies;
+    std::vector<std::tuple<edge_t*, edge_t const*, size_t>> edge_id_copies;
+    std::vector<std::tuple<edge_type_t*, edge_type_t const*, size_t>> edge_type_copies;
+
+    {
+      std::lock_guard<std::mutex> lock(lock_);
+
+      size_t count = src.size();
+      size_t pos   = 0;
+
+      while (count > 0) {
+        size_t copy_count = std::min(count, (src_.back().size() - current_pos_));
+
+        src_copies.push_back(
+          std::make_tuple(src_.back().begin() + current_pos_, src.begin() + pos, copy_count));
+        dst_copies.push_back(
+          std::make_tuple(dst_.back().begin() + current_pos_, dst.begin() + pos, copy_count));
+        if (wgt)
+          wgt_copies.push_back(
+            std::make_tuple(wgt_->back().begin() + current_pos_, wgt->begin() + pos, copy_count));
+        if (edge_id)
+          edge_id_copies.push_back(std::make_tuple(
+            edge_id_->back().begin() + current_pos_, edge_id->begin() + pos, copy_count));
+        if (edge_type)
+          edge_type_copies.push_back(std::make_tuple(
+            edge_type_->back().begin() + current_pos_, edge_type->begin() + pos, copy_count));
+
+        count -= copy_count;
+        pos += copy_count;
+        current_pos_ += copy_count;
+      }
```

Comment:
What happens if `count` = 1000, `src_.back().size()` = 100, and `current_pos_` = 0?

At the end of the first loop, `copy_count` = 100, `count` = 900, `pos`=100, `current_pos_`=100. From the second loop, `copy_count`=0 and this loop won't finish or am I missing something?

Shouldn't we allocate additional buffers and reset `current_pos_` to 0 for this loop to finish?